### PR TITLE
Remove unused output

### PR DESCRIPTION
**Pull request by joaander**
## Description

Remove code that outputs unnecessary to stdout/stderr.

* Remove `hoomd.cite`: #748
* Replace build information print with variables in `hoomd.version`
* Remove `printStats`

## Motivation and context

One of the design guidelines for HOOMD v3 is to behave as a Python package and not provide unnecessary output by default.

Much of the information removed in `printStats` is (or can be made) available via operation's properties, allowing the user to request the information at a granular level and in a machine readable format when needed.

Resolves: #684

## How has this been tested?

* Unit tests still pass. 
* Example scripts do not produce output.

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by joaander** [comment]
I also added a `write_debug_data` to capture information useful for troubleshooting issues. Here is some sample output from an HPMC simulation:
```
{
    "hoomd_module": "<module 'hoomd' from '/scratch/joaander/build/hoomd/hoomd/__init__.py'>",
    "version": {
        "compile_date": "2020-09-11",
        "compile_flags": "GPU [CUDA] (11.0) DOUBLE HPMC_MIXED SSE SSE2 SSE3 SSE4_1 SSE4_2 AVX AVX2 ",
        "cxx_compiler": "gcc 9.3.0",
        "enable_gpu": true,
        "enable_mpi": false,
        "enable_tbb": false,
        "git_branch": "refs/heads/remove-unused-output",
        "git_sha1": "5f4c7e7fe6e0d43159978c3d2c3ca0b414741f67",
        "gpu_api_version": "11.0",
        "gpu_platform": "CUDA",
        "install_dir": "/lib/python3.8/site-packages/hoomd",
        "source_dir": "/scratch/joaander/devel/hoomd"
    },
    "device": {
        "msg_file": null,
        "notice_level": 2,
        "devices": [
            "CPU"
        ],
        "num_cpu_threads": 1,
        "gpu_available_devices": [
            "[0]       Quadro RTX 5000  48 SM_7.5 @ 1.82 GHz, 16124 MiB DRAM"
        ],
        "gpu_unavailable_device_reasons": []
    },
    "communicator": {
        "num_ranks": 1
    },
    "state": {
        "types": {
            "particle_types": [
                "octahedron"
            ],
            "bond_types": [],
            "angle_types": [],
            "dihedral_types": [],
            "improper_types": [],
            "special_pair_types": []
        },
        "N_particles": 250,
        "N_bonds": 0,
        "N_angles": 0,
        "N_impropers": 0,
        "N_special_pairs": 0,
        "N_dihedrals": 0,
        "box": "hoomd.box.Box(Lx=7.699999809265137, Ly=7.699999809265137, Lz=7.699999809265137, xy=0.0, xz=0.0, yz=0.0)"
    },
    "operations": {
        "simulation": {
            "Simulation": {
                "timestep": 10000
            }
        },
        "hpmc": {
            "integrate": {
                "ConvexPolyhedron": {
                    "state": {
                        "d": {
                            "octahedron": 0.1,
                            "__default__": 0.1
                        },
                        "a": {
                            "octahedron": 0.1,
                            "__default__": 0.1
                        },
                        "depletant_fugacity": {
                            "octahedron": 0.0,
                            "__default__": 0.0
                        },
                        "interaction_matrix": {
                            "('octahedron', 'octahedron')": true,
                            "__default__": true
                        },
                        "shape": {
                            "octahedron": {
                                "vertices": [
                                    [
                                        -0.5,
                                        0.0,
                                        0.0
                                    ],
                                    [
                                        0.5,
                                        0.0,
                                        0.0
                                    ],
                                    [
                                        0.0,
                                        -0.5,
                                        0.0
                                    ],
                                    [
                                        0.0,
                                        0.5,
                                        0.0
                                    ],
                                    [
                                        0.0,
                                        0.0,
                                        -0.5
                                    ],
                                    [
                                        0.0,
                                        0.0,
                                        0.5
                                    ]
                                ],
                                "ignore_statistics": 0,
                                "sweep_radius": 0.0
                            },
                            "__default__": {
                                "vertices": [
                                    [
                                        "<class 'hoomd.typeconverter.RequiredArg'>",
                                        "<class 'hoomd.typeconverter.RequiredArg'>",
                                        "<class 'hoomd.typeconverter.RequiredArg'>"
                                    ]
                                ],
                                "sweep_radius": 0.0,
                                "ignore_statistics": false
                            }
                        },
                        "__params__": {
                            "seed": 2,
                            "move_ratio": 0.5,
                            "nselect": 4
                        }
                    },
                    "map_overlaps": [],
                    "overlaps": 0,
                    "translate_moves": [
                        4487025,
                        510892
                    ],
                    "rotate_moves": [
                        4753854,
                        248229
                    ],
                    "mps": 694844.5108221684,
                    "type_shapes": [
                        {
                            "type": "ConvexPolyhedron",
                            "rounding_radius": 0,
                            "vertices": [
                                [
                                    -0.5,
                                    0,
                                    0
                                ],
                                [
                                    0.5,
                                    0,
                                    0
                                ],
                                [
                                    0,
                                    -0.5,
                                    0
                                ],
                                [
                                    0,
                                    0.5,
                                    0
                                ],
                                [
                                    0,
                                    0,
                                    -0.5
                                ],
                                [
                                    0,
                                    0,
                                    0.5
                                ]
                            ]
                        }
                    ]
                }
            }
        },
        "tune": {
            "sorter": {
                "ParticleSorter": {
                    "state": {
                        "__params__": {
                            "trigger": "<hoomd.trigger.Periodic object at 0x7f45a8bf96d0>",
                            "grid": 256
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by joaander** [comment]
Here is a sample from an MD simulation with two pair forces:
```
{
    "hoomd_module": "<module 'hoomd' from '/scratch/joaander/build/hoomd/hoomd/__init__.py'>",
    "version": {
        "compile_date": "2020-09-14",
        "compile_flags": "GPU [CUDA] (11.0) DOUBLE HPMC_MIXED SSE SSE2 SSE3 SSE4_1 SSE4_2 AVX AVX2 ",
        "cxx_compiler": "gcc 9.3.0",
        "enable_gpu": true,
        "enable_mpi": false,
        "enable_tbb": false,
        "git_branch": "refs/heads/remove-unused-output",
        "git_sha1": "6f5cbeb7b47c062942ba116b2522936794fab830",
        "gpu_api_version": "11.0",
        "gpu_platform": "CUDA",
        "install_dir": "/lib/python3.8/site-packages/hoomd",
        "source_dir": "/scratch/joaander/devel/hoomd"
    },
    "device": {
        "msg_file": null,
        "notice_level": 2,
        "devices": [
            "CPU"
        ],
        "num_cpu_threads": 1,
        "gpu_available_devices": [
            "[0]       Quadro RTX 5000  48 SM_7.5 @ 1.82 GHz, 16124 MiB DRAM"
        ],
        "gpu_unavailable_device_reasons": []
    },
    "communicator": {
        "num_ranks": 1
    },
    "state": {
        "types": {
            "particle_types": [
                "A"
            ],
            "bond_types": [],
            "angle_types": [],
            "dihedral_types": [],
            "improper_types": [],
            "special_pair_types": []
        },
        "N_particles": 2000,
        "N_bonds": 0,
        "N_angles": 0,
        "N_impropers": 0,
        "N_special_pairs": 0,
        "N_dihedrals": 0,
        "box": "hoomd.box.Box(Lx=15.399999618530273, Ly=15.399999618530273, Lz=15.399999618530273, xy=0.0, xz=0.0, yz=0.0)"
    },
    "operations": {
        "simulation": {
            "Simulation": {
                "timestep": 10
            }
        },
        "md": {
            "integrate": {
                "Integrator": {
                    "state": {
                        "__params__": {
                            "dt": 0.005,
                            "aniso": "auto"
                        }
                    }
                }
            },
            "pair": {
                "LJ": {
                    "state": {
                        "r_cut": {
                            "('A', 'A')": 2.5
                        },
                        "r_on": {
                            "('A', 'A')": 0.0,
                            "__default__": 0.0
                        },
                        "params": {
                            "('A', 'A')": {
                                "sigma": 1.2,
                                "epsilon": 1.0
                            }
                        },
                        "__params__": {
                            "mode": "none"
                        }
                    },
                    "energy": 12443.395507328998,
                    "energies": "[ 5.55903041  6.13302026  5.59196472 ...  1.46921199  0.08423068\n -1.28992816]",
                    "forces": "[[-26.84492497   6.6150191    2.60357503]\n [-24.05909092   0.33425672   0.81359081]\n [-32.94299103   9.40393183   6.80699647]\n ...\n [ 24.24120923   3.71989292   4.79495259]\n [ 15.72671947  -1.53794807  29.76036284]\n [ 25.47643272   0.23615341  40.52529382]]",
                    "torques": "[[0. 0. 0.]\n [0. 0. 0.]\n [0. 0. 0.]\n ...\n [0. 0. 0.]\n [0. 0. 0.]\n [0. 0. 0.]]",
                    "virials": null
                },
                "Gauss": {
                    "state": {
                        "r_cut": {
                            "('A', 'A')": 3.5
                        },
                        "r_on": {
                            "('A', 'A')": 0.0,
                            "__default__": 0.0
                        },
                        "params": {
                            "('A', 'A')": {
                                "sigma": 0.9,
                                "epsilon": 1.5
                            }
                        },
                        "__params__": {
                            "mode": "none"
                        }
                    },
                    "energy": 9619.177707531,
                    "energies": "[3.98296084 4.01797463 4.03353279 ... 2.55276226 2.49849311 2.21615032]",
                    "forces": "[[-4.10301705 -0.00959827 -0.09257686]\n [-4.01682542  0.00898229 -0.04257   ]\n [-3.98974091  0.02141377  0.02247802]\n ...\n [ 3.42258351  0.06821831  0.01796346]\n [ 3.4256924  -0.01507294  0.21511529]\n [ 3.56541337 -0.00426207  1.04322612]]",
                    "torques": "[[0. 0. 0.]\n [0. 0. 0.]\n [0. 0. 0.]\n ...\n [0. 0. 0.]\n [0. 0. 0.]\n [0. 0. 0.]]",
                    "virials": null
                }
            },
            "methods": {
                "Langevin": {
                    "state": {
                        "gamma": {
                            "A": 1.0,
                            "__default__": 1.0
                        },
                        "gamma_r": {
                            "A": [
                                1.0,
                                1.0,
                                1.0
                            ],
                            "__default__": [
                                1.0,
                                1.0,
                                1.0
                            ]
                        },
                        "__params__": {
                            "filter": "ParticleFilter.All",
                            "kT": 1.5,
                            "seed": 6,
                            "alpha": null,
                            "tally_reservoir_energy": false
                        }
                    }
                }
            },
            "nlist": {
                "Cell": {
                    "state": {
                        "__params__": {
                            "exclusions": [
                                "bond"
                            ],
                            "buffer": 0.4,
                            "rebuild_check_delay": 1,
                            "check_dist": true,
                            "diameter_shift": false,
                            "max_diameter": 1.0,
                            "deterministic": false
                        }
                    },
                    "shortest_rebuild": 99
                }
            }
        },
        "tune": {
            "sorter": {
                "ParticleSorter": {
                    "state": {
                        "__params__": {
                            "trigger": "hoomd.trigger.Periodic(period=200, phase=0",
                            "grid": 256
                        }
                    }
                }
            }
        }
    }
}```

File: hoomd/trigger.py [hunk 1]
```diff
@@ -14,6 +14,10 @@ class Periodic(_hoomd.PeriodicTrigger, Trigger):
     def __init__(self, period, phase=0):
         _hoomd.PeriodicTrigger.__init__(self, period, phase)
 
+    def __str__(self):
+        return f"hoomd.trigger.Periodic(period={self.period}, " \
+               f"phase={self.phase}"
```
Comment (b-butler) [inline hoomd/trigger.py:19]:
```suggestion
               f"phase={self.phase})"
```

File: hoomd/trigger.py [hunk 2]
```diff
@@ -22,6 +26,8 @@ def __init__(self, timestep):
         else:
             _hoomd.BeforeTrigger.__init__(self, timestep)
 
+    def __str__(self):
+        return f"hoomd.trigger.Before(timestep={self.timestep}"
```
Comment (b-butler) [inline hoomd/trigger.py:30]:
```suggestion
        return f"hoomd.trigger.Before(timestep={self.timestep})"
```

File: hoomd/trigger.py [hunk 3]
```diff
@@ -30,6 +36,8 @@ def __init__(self, timestep):
         else:
             _hoomd.OnTrigger.__init__(self, timestep)
 
+    def __str__(self):
+        return f"hoomd.trigger.On(timestep={self.timestep}"
```
Comment (b-butler) [inline hoomd/trigger.py:40]:
```suggestion
        return f"hoomd.trigger.On(timestep={self.timestep})"
```

File: hoomd/trigger.py [hunk 4]
```diff
@@ -38,11 +46,15 @@ def __init__(self, timestep):
         else:
             _hoomd.AfterTrigger.__init__(self, timestep)
 
+    def __str__(self):
+        return f"hoomd.trigger.After(timestep={self.timestep}"
```
Comment (b-butler) [inline hoomd/trigger.py:50]:
```suggestion
        return f"hoomd.trigger.After(timestep={self.timestep})"
```

File: hoomd/trigger.py [hunk 5]
```diff
@@ -38,11 +46,15 @@ def __init__(self, timestep):
...
 
 class Not(_hoomd.NotTrigger, Trigger):
     def __init__(self, trigger):
         _hoomd.NotTrigger.__init__(self, trigger)
 
+    def __str__(self):
+        return f"hoomd.trigger.Not(timestep={self.timestep}"
```
Comment (b-butler) [inline hoomd/trigger.py:57]:
```suggestion
        return f"hoomd.trigger.Not(timestep={self.timestep})"
```

File: hoomd/trigger.py [hunk 6]
```diff
@@ -38,11 +46,15 @@ def __init__(self, timestep):
...
 
 class And(_hoomd.AndTrigger, Trigger):
```
Comment (b-butler) [inline hoomd/trigger.py:59]: Should we give `And` and `Or` triggers a `__str__` method that does something like `hoomd.trigger.And([...])` where the ellipisis are replaced with the return string of `str(trigger)`?

File: hoomd/version.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    gpu_platform (str): Name of the GPU platform this build was compiled
+        against.
+
+    install_dir (str): The installation directory.
+
+    source_dir (str): The source directory.
+"""
+from hoomd import _hoomd
+
+version = "2.9.2"
```
Comment (b-butler) [inline hoomd/version.py:36]: Even though it isn't PEP8 compliant, I feel we should probably follow the tradition of using `__version__` as that is what most people will expect.

**Review by b-butler — CHANGES_REQUESTED** [review]
Thanks for all the work. Overall it looks great. I have left a few comments. I think some of the `BuildInfo` and `version` methods/attributes would be better named (in Python at least) `mpi_enabled` rather than the action `enable_mpi`.

File: hoomd/operation.py [hunk 8]
```diff
@@ -452,7 +452,13 @@ def _update_param_dict(self):
 
     @log(flag='state')
     def state(self):
-        self._update_param_dict()
+        try:
+            self._update_param_dict()
+        except AttributeError:
+            # attribute errors in _update_param_dict should not result in
+            # class has no attribute `state`
+            raise RuntimeError("Failed to update param dict")
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/operation.py:460]: This change converts this error:
```
Traceback (most recent call last):
  File "test2.py", line 29, in <module>
    sim.write_debug_data('test.json')
  File "/scratch/joaander/build/hoomd/hoomd/simulation.py", line 248, in write_debug_data
    log = logger.log()
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in log
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 49, in dict_map
    new_dict[key] = func(value)
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in <lambda>
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 362, in __call__
    attr = getattr(self.obj, self.attr)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 94, in __getattr__
    raise AttributeError("Object {} has no attribute {}".format(
AttributeError: Object <class 'hoomd.md.nlist.Cell'> has no attribute state

```

Into this one that identifies the root cause:
```
Traceback (most recent call last):
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 456, in state
    self._update_param_dict()
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 451, in _update_param_dict
    self._param_dict[key] = getattr(self, key)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 90, in __getattr__
    return self._getattr_param(attr)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 377, in _getattr_param
    return getattr(self._cpp_obj, attr)
AttributeError: 'hoomd.md._md.NeighborListBinned' object has no attribute 'deterministic'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test2.py", line 29, in <module>
    sim.write_debug_data('test.json')
  File "/scratch/joaander/build/hoomd/hoomd/simulation.py", line 248, in write_debug_data
    log = logger.log()
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in log
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 47, in dict_map
    new_dict[key] = dict_map(value, func)
  File "/scratch/joaander/build/hoomd/hoomd/util.py", line 49, in dict_map
    new_dict[key] = func(value)
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 636, in <lambda>
    return dict_map(self._dict, lambda x: x())
  File "/scratch/joaander/build/hoomd/hoomd/logging.py", line 362, in __call__
    attr = getattr(self.obj, self.attr)
  File "/scratch/joaander/build/hoomd/hoomd/operation.py", line 459, in state
    raise RuntimeError("Failed to update param dict")
RuntimeError: Failed to update param dict
```

Without this change, the `AttributeError` thrown by `_md.NeighborListBinned` causes Python to assume that `Cell` has no attribute `state`. I don't know if there is a more Pythonic way to do this.

I would guess that any `AttributeError` thrown in `_get_state` would also result in the same behavior so that call should possibly also be inside the try: block.

File: hoomd/operation.py [hunk 9]
```diff
@@ -466,6 +472,15 @@ def _extend_typeparam(self, typeparams):
...
             self._add_typeparam(typeparam)
 
+    @property
+    def _children(self):
+        """A set of child objects.
+
+        These objects do not appear directly in any of the operations lists but
+        are owned in lists or members of those operations.
+        """
+        return set()
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/operation.py:482]: fixed.

File: hoomd/simulation.py [hunk 10]
```diff
@@ -172,27 +161,96 @@ def always_compute_pressure(self, value):
...
+                box=repr(self.state.box))
+
+        # save all loggable quantities from operations and their child computes
+        logger = hoomd.logging.Logger(only_default=False)
+        logger += self
+
+        # children may appear several times, identify them uniquely
+        children = list()
+        for op in self.operations:
+            if op is not None:
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/simulation.py:240]: done

File: hoomd/simulation.py [hunk 11]
```diff
@@ -172,27 +161,96 @@ def always_compute_pressure(self, value):
...
+                logger.add(op)
+
+                for child in op._children:
+                    if child not in children:
+                        children.append(child)
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/simulation.py:245]: It does not appear multiple times, but I do not know why.

File: hoomd/simulation.py [hunk 12]
```diff
@@ -172,27 +161,96 @@ def always_compute_pressure(self, value):
...
+                N_particles=self.state.N_particles,
+                N_bonds=self.state.N_bonds,
+                N_angles=self.state.N_angles,
+                N_impropers=self.state.N_impropers,
+                N_special_pairs=self.state.N_special_pairs,
+                N_dihedrals=self.state.N_dihedrals,
+                box=repr(self.state.box))
+
+        # save all loggable quantities from operations and their child computes
+        logger = hoomd.logging.Logger(only_default=False)
```
[the remark this answers is not in the document]
Comment (joaander) [reply hoomd/simulation.py:234]: We can add logic to handle any special cases as we find them. For now, the `str()` default conversion relies on `numpy`'s array->string conversion method which summarizes long arrays (ex: `"energies": "[3.98296084 4.01797463 4.03353279 ... 2.55276226 2.49849311 2.21615032]"`)

File: hoomd/trigger.py (hunk 1 above)
Comment (joaander) [reply hoomd/trigger.py:19]: done

File: hoomd/trigger.py (hunk 2 above)
Comment (joaander) [reply hoomd/trigger.py:30]: done

File: hoomd/trigger.py (hunk 6 above)
Comment (joaander) [reply hoomd/trigger.py:59]: done

Comment (joaander) [reply hoomd/trigger.py:59]: Unfortunately, the objects accessed from the `triggers` list don't get the `__str__` method defined in Python:
```
'hoomd.trigger.And([<hoomd._hoomd.BeforeTrigger object at 0x7f8a29704d70>, <hoomd._hoomd.PeriodicTrigger object at 0x7f89f9da18b0>])'
```
We could solve this by implementing `__str__` in C++, but I'm not sure that it is worth it at this time.

File: hoomd/version.py (hunk 7 above)
Comment (joaander) [reply hoomd/version.py:36]: I disagree. Not only is it not PEP8 compliant, `__*__` are defined as system-defined names in the Python language specification: https://docs.python.org/3/reference/lexical_analysis.html#reserved-classes-of-identifiers
```
System-defined names, informally known as “dunder” names. These names are defined by the interpreter and its implementation (including the standard library). Current system names are discussed in the Special method names section and elsewhere. More will likely be defined in future versions of Python. Any use of __*__ names, in any context, that does not follow explicitly documented use, is subject to breakage without warning.
```
`__version__`, while commonly used, is not universal and it is not documented in the Python language specification or documentation: https://docs.python.org/3/search.html?q=__version__&check_keywords=yes&area=default
Users should expect to find the HOOMD-blue version in the attribute we document in the sphinx docs.

In the Python standard library, for example, a few packages provide `__version__`, (such as `http.server` and `json`), however these are being removed over time (for example, see the Python 3.5 release notes https://docs.python.org/3/whatsnew/3.5.html where `__version__` was dropped from the `email` package). Of these, only `http.server.__version__` is actually documented. Interestingly, `curses.__version__` is documented, but doesn't exist in Python 3.8. `curses`, `ensurepip`, `marshal`, and others provide a `version` attribute, function (which sometimes provides a string and sometimes a tuple).

The `version` module in HOOMD follows that of `numpy` e.g. `numpy.version.git_revision`

Comment (b-butler) [reply hoomd/version.py:36]: Okay, well I am not opinionated enough to argue against this.

File: hoomd/simulation.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/simulation.py:234]: I was wondering whether that would result in lists or strings. I saw the `default` conversion, but didn't know if the `json` module would handle iterables differently. That works then.

File: hoomd/simulation.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/simulation.py]: When adding objects to the logger we check if that exact object has been logged in that namespace before (i.e. `obj1 is obj2`). Since the neighbor list will always return the same namespace, we see it is already in use and that it is in use by the same object and so perform a no-op.

File: hoomd/simulation.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/simulation.py]: Awesome.

File: hoomd/operation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (b-butler) [reply hoomd/operation.py:460]: This is fine. I have just noticed that properties can mask errors, and I hadn't found a good way to prevent that. I like this approach.

File: hoomd/trigger.py (hunk 6 above)
Comment (b-butler) [reply hoomd/trigger.py:59]: I wouldn't bother for this PR. It would be worth doing later though.

File: hoomd/simulation.py [hunk 13]
```diff
@@ -237,12 +237,11 @@ def write_debug_data(self, filename):
...
-            if op is not None:
-                logger.add(op)
+            logger.add(op)
 
-                for child in op._children:
-                    if child not in children:
-                        children.append(child)
+            for child in op._children:
+                if child not in children:
+                    children.append(child)
```
Comment (b-butler) [inline hoomd/simulation.py:244]: This should be unnecessary as I explain in another comment.

**Review by b-butler — COMMENTED** [review]
I only have a couple of comments. I like that we are testing the `__str__` methods of the trigger, but it should really be done in its own separate function. I get the convenience of doing it in the eval test function though. We can just put an issue for now so this gets tracked, and move on with this PR though.

Also, I just want to say that I think the `hoomd.version` module attributes are really nice and Pythonic. This makes the Python code for checking `gpu`, `mpi`, and other compatibility and just general build information much more readable, great work. The JSON debug output will also be really nice and help a lot with personal debugging and answering others questions.

**Comment by joaander** [comment]
The same AttributeError issue occurs with properties like HPMC's `translate_moves` before attaching:
```
def translate_moves(self):
    return self._cpp_obj.getCounters(1).translate
```
```
AttributeError: Object <class 'hoomd.hpmc.integrate.ConvexPolyhedron'> has no attribute translate_moves
```
There must be a better solution than to wrap every property's body in a try/except block that raises a different exception.

Also, we should make a task issue to test that every class can log all quantities before and after attaching.

**Comment by b-butler** [comment]
I agree with the loggable quantity thing. We could honestly implement a check function in `conftest` that we just use everywhere to test that all loggable quantities are valid before and after attaching.

I honestly, don't know a better way to handle properties than to "code well". If the attribute error doesn't get raised, we don't get the problem, but I don't know how to prevent this from happening in general.

File: hoomd/HOOMDVersion.h.in [hunk 14]
```diff
@@ -7,10 +7,6 @@
 
 // Full text HOOMD version string
 #define HOOMD_VERSION "${HOOMD_VERSION}"
-// Full text HOOMD version string
-#define HOOMD_VERSION_LONG "${HOOMD_VERSION_LONG}"
-// Text string containing the date this build was configured
-#define COMPILE_DATE "${COMPILE_DATE}"
```
Comment (b-butler) [inline hoomd/HOOMDVersion.h.in:12]: Why are we moving this from C++ to Python. I don't have a preference necessarily, but wondering why the change was made.

Comment (joaander) [reply hoomd/HOOMDVersion.h.in:12]: We don't use these in C++ anywhere important so they don't need to be there. Since `COMPILE_DATE` changes every day and `HOOMD_VERSION_LONG` includes the git SHA, these cause the header to change regularly and make for a number of needless recompiles and relinks. Since I am refactoring this file anyways, I thought to solve that issue. It does cause some difficulty for Sphinx docs, though so maybe I will just revert the change.
